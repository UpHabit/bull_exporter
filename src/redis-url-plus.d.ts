declare module 'redis-url-plus' {
    export default function redisUrlPlus(url: string): import('ioredis').RedisOptions;
}