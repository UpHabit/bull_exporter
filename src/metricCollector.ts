import bull from 'bull';
import * as Logger from 'bunyan';
import { EventEmitter } from 'events';
import IoRedis from 'ioredis';
import { register as globalRegister, Registry } from 'prom-client';
import URL from 'url';

import { logger as globalLogger } from './logger';
import { getJobCompleteStats, getStats, makeGuages, QueueGauges } from './queueGauges';

export interface MetricCollectorOptions extends Omit<bull.QueueOptions, 'redis'> {
  metricPrefix: string;
  redis: string;
  sentinelExtraEndpoints: string[];
  sentinelMasterName: string;
  autoDiscover: boolean;
  logger: Logger;
}

export interface QueueData<T = unknown> {
  queue: bull.Queue<T>;
  name: string;
  prefix: string;
}

export class MetricCollector {

  private readonly logger: Logger;

  private readonly defaultRedisClient: IoRedis.Redis;
  private readonly redisConnectionCredentials: string | IoRedis.RedisOptions;
  private readonly bullOpts: Omit<bull.QueueOptions, 'redis'>;
  private readonly queuesByName: Map<string, QueueData<unknown>> = new Map();

  private get queues(): QueueData<unknown>[] {
    return [...this.queuesByName.values()];
  }

  private readonly myListeners: Set<(id: string) => Promise<void>> = new Set();

  private readonly guages: QueueGauges;

  private static parseSentinelUri(redisUri: string, sentinelExtraEndpoints: string[], sentinelMasterName?: string): string | IoRedis.RedisOptions {
    const { protocol, host, port, auth } = URL.parse(redisUri);

    if (protocol !== 'sentinel:') {
      return redisUri;
    }

    const password = auth ? auth.split(':')[1] : undefined;

    if (host === undefined || port === undefined || sentinelExtraEndpoints === undefined) {
      throw new Error('');
    }

    return {
      password,
      sentinels: [{ host, port: Number(port) }].concat(sentinelExtraEndpoints.map(sentinelExtraEndpoint => {
        const [additionalHost, additionalPort] = sentinelExtraEndpoint.split(':');
        return { host: additionalHost, port: Number(additionalPort) };
      })),
      name: sentinelMasterName,
    };
  }
  constructor(
    queueNames: string[],
    opts: MetricCollectorOptions,
    registers: Registry[] = [globalRegister],
  ) {
    const { logger, autoDiscover, redis, metricPrefix, sentinelExtraEndpoints, sentinelMasterName, ...bullOpts } = opts;
    this.redisConnectionCredentials = MetricCollector.parseSentinelUri(redis, sentinelExtraEndpoints, sentinelMasterName === '' ? undefined : sentinelMasterName);
    if (typeof this.redisConnectionCredentials === 'string') {
      this.defaultRedisClient = new IoRedis(this.redisConnectionCredentials);
    } else {
      this.defaultRedisClient = new IoRedis(this.redisConnectionCredentials);
    }
    this.defaultRedisClient.setMaxListeners(32);
    this.bullOpts = bullOpts;
    this.logger = logger || globalLogger;
    this.addToQueueSet(queueNames);
    this.guages = makeGuages(metricPrefix, registers);
  }

  private createClient(_type: 'client' | 'subscriber' | 'bclient', redisOpts?: IoRedis.RedisOptions): IoRedis.Redis {
    if (_type === 'client') {
      return this.defaultRedisClient!;
    }
    if (typeof this.redisConnectionCredentials === 'string') {
      return new IoRedis(this.redisConnectionCredentials, redisOpts);
    }
    return new IoRedis({ ...this.redisConnectionCredentials, ...redisOpts });
  }

  private addToQueueSet(names: string[]): void {
    for (const name of names) {
      if (this.queuesByName.has(name)) {
        continue;
      }
      this.logger.info('added queue', name);
      this.queuesByName.set(name, {
        name,
        queue: new bull(name, {
          ...this.bullOpts,
          createClient: this.createClient.bind(this),
        }),
        prefix: this.bullOpts.prefix || 'bull',
      });
    }
  }

  public async discoverAll(): Promise<void> {
    const keyPattern = new RegExp(`^${this.bullOpts.prefix}:([^:]+):(id|failed|active|waiting|stalled-check)$`);
    this.logger.info({ pattern: keyPattern.source }, 'running queue discovery');

    const keyStream = this.defaultRedisClient.scanStream({
      match: `${this.bullOpts.prefix}:*:*`,
    });
    // tslint:disable-next-line:await-promise tslint does not like Readable's here
    for await (const keyChunk of keyStream) {
      for (const key of keyChunk) {
        const match = keyPattern.exec(key);
        if (match && match[1]) {
          this.addToQueueSet([match[1]]);
        }
      }
    }
  }

  private async onJobComplete(queue: QueueData, id: string): Promise<void> {
    try {
      const job = await queue.queue.getJob(id);
      if (!job) {
        this.logger.warn({ job: id }, 'unable to find job from id');
        return;
      }
      await getJobCompleteStats(queue.prefix, queue.name, job, this.guages);
    } catch (err) {
      this.logger.error({ err, job: id }, 'unable to fetch completed job');
    }
  }

  public collectJobCompletions(): void {
    for (const q of this.queues) {
      const cb = this.onJobComplete.bind(this, q);
      this.myListeners.add(cb);
      q.queue.on('global:completed', cb);
    }
  }

  public async updateAll(): Promise<void> {
    const updatePromises = this.queues.map(q => getStats(q.prefix, q.name, q.queue, this.guages));
    await Promise.all(updatePromises);
  }

  public async ping(): Promise<void> {
    await this.defaultRedisClient.ping();
  }

  public async close(): Promise<void> {
    this.defaultRedisClient.disconnect();
    for (const q of this.queues) {
      for (const l of this.myListeners) {
        (q.queue as any as EventEmitter).removeListener('global:completed', l);
      }
    }
    await Promise.all(this.queues.map(q => q.queue.close()));
  }

}
