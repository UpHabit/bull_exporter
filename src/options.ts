import yargs from 'yargs';

import { version } from '../package.json';

export interface Options {
  url: string;
  sentinelExtraEndpoints: string[];
  sentinelMasterName: string;
  prefix: string;
  metricPrefix: string;
  once: boolean;
  port: number;
  bindAddress: string;
  autoDiscover: boolean;
  _: string[];
}

export function getOptionsFromArgs(...args: string[]): Options {
  return yargs
    .version(version)
    .alias('V', 'version')
    .options({
      url: {
        alias: 'u',
        describe: 'A redis connection url. To connect to Redis Sentinel use "sentinel:" schema',
        default: 'redis://127.0.0.1:6379',
        demandOption: true,
      },
      sentinelExtraEndpoints: {
        describe: 'Extra host:port pairs for Redis Sentinel',
        default: [],
        array: true,
        demandOption: true,
      },
      sentinelMasterName: {
        describe: 'Redis Sentinel Master Name',
        default: '',
        demandOption: true,
      },
      prefix: {
        alias: 'p',
        default: 'bull',
        demandOption: true,
      },
      metricPrefix: {
        alias: 'm',
        default: 'bull_queue_',
        defaultDescription: 'prefix for all exported metrics',
        demandOption: true,
      },
      once: {
        alias: 'n',
        default: false,
        type: 'boolean',
        description: 'Print stats and exit without starting a server',
      },
      port: {
        default: 9538,
      },
      autoDiscover: {
        default: false,
        alias: 'a',
        type: 'boolean',
      },
      bindAddress: {
        alias: 'b',
        description: 'Address to listen on',
        default: '0.0.0.0',
      },
    }).parse(args);
}
