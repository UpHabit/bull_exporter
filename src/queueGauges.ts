import { Job, Queue } from 'bullmq';
import { Gauge, Registry, Summary } from 'prom-client';

type LabelsT = 'queue' | 'prefix';
export interface QueueGauges {
	completed: Gauge<LabelsT>;
	active: Gauge<LabelsT>;
	delayed: Gauge<LabelsT>;
	failed: Gauge<LabelsT>;
	waiting: Gauge<LabelsT>;
	completeSummary: Summary<LabelsT>;
	prioritized: Gauge<LabelsT>;
}

export function makeGuages(statPrefix: string, registers: Registry[]): QueueGauges {
	return {
		completed: new Gauge({
			registers,
			name: `${statPrefix}completed`,
			help: 'Number of completed messages',
			labelNames: ['queue', 'prefix'],
		}),
		completeSummary: new Summary({
			registers,
			name: `${statPrefix}complete_duration`,
			help: 'Time to complete jobs',
			labelNames: ['queue', 'prefix'],
			maxAgeSeconds: 300,
			ageBuckets: 13,
		}),
		active: new Gauge({
			registers,
			name: `${statPrefix}active`,
			help: 'Number of active messages',
			labelNames: ['queue', 'prefix'],
		}),
		delayed: new Gauge({
			registers,
			name: `${statPrefix}delayed`,
			help: 'Number of delayed messages',
			labelNames: ['queue', 'prefix'],
		}),
		failed: new Gauge({
			registers,
			name: `${statPrefix}failed`,
			help: 'Number of failed messages',
			labelNames: ['queue', 'prefix'],
		}),
		waiting: new Gauge({
			registers,
			name: `${statPrefix}waiting`,
			help: 'Number of waiting messages',
			labelNames: ['queue', 'prefix'],
		}),
		prioritized: new Gauge({
			registers,
			name: `${statPrefix}prioritized`,
			help: 'Number of prioritized messages',
			labelNames: ['queue', 'prefix'],
		}),
	};
}

export async function getJobCompleteStats(prefix: string, name: string, job: Job, gauges: QueueGauges): Promise<void> {
	if (!job.finishedOn) {
		return;
	}
	const duration = job.finishedOn - job.processedOn!;
	gauges.completeSummary.observe({ prefix, queue: name }, duration);
}

export async function getStats(prefix: string, name: string, queue: Queue, gauges: QueueGauges): Promise<void> {
	const { completed, active, delayed, failed, waiting, prioritized } = await queue.getJobCounts();

	gauges.completed.set({ prefix, queue: name }, completed);
	gauges.active.set({ prefix, queue: name }, active);
	gauges.delayed.set({ prefix, queue: name }, delayed);
	gauges.failed.set({ prefix, queue: name }, failed);
	gauges.waiting.set({ prefix, queue: name }, waiting);
	gauges.prioritized.set({ prefix, queue: name }, prioritized);
}
