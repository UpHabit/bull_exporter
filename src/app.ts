import express from 'express';
import * as http from 'http';

import { logger } from './utils/logger';
import collectorRoutes from './collector/routes';
import { getOptions } from './utils/options';
import { handleError } from './middlewares/error-handler-middleware';
import { setHeaders } from './middlewares/headers-middleware';
import { logRequestDetails } from './middlewares/request-stats-middleware';
import CollectorApi from './collector/api';

export async function makeServer(): Promise<express.Application> {
	const app = express();
	app.disable('x-powered-by');

	app.use(setHeaders);
	app.use(logRequestDetails);

	await CollectorApi.startCollector();

	app.use('/', collectorRoutes);

	app.use(handleError);

	return app;
}

export async function startServer(): Promise<{ done: Promise<void> }> {
	const app = await makeServer();

	let server: http.Server;
	const opts = getOptions();
	await new Promise<void>((resolve) => {
		server = app.listen(opts.port, opts.bindAddress, () => {
			logger.info(`Running on ${opts.bindAddress}:${opts.port}`);
			resolve();
		});
	});

	process.on('SIGTERM', () => server.close());

	const done = new Promise<void>((resolve, reject) => {
		server.on('close', () => resolve());
		server.on('error', (err: any) => reject(err));
	});

	return { done };
}
