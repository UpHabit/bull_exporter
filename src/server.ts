import express from 'express';
import * as http from 'http';
import promClient from 'prom-client';
import { v4 as uuid } from 'uuid';

import { logger } from './logger';
import { MetricCollector } from './metricCollector';
import { Options } from './options';

function calcDuration(start: [number, number]): number {
  const diff = process.hrtime(start);
  return diff[0] * 1e3 + diff[1] * 1e-6;
}

export async function makeServer(opts: Options): Promise<express.Application> {
  const app = express();
  app.disable('x-powered-by');

  app.use((_req: express.Request, res: express.Response, next: express.NextFunction) => {
    res.header('Content-Security-Policy', `default-src 'none'; form-action 'none'`);
    res.header('X-Permitted-Cross-Domain-Policies', 'none');
    res.header('Pragma', 'no-cache');
    res.header('Cache-Control', 'no-store, no-cache, must-revalidate, proxy-revalidate');
    res.header('Content-Type-Options', 'nosniff');
    res.header('XSS-Protection', '1; mode=block');
    next();
  });

  app.use((req: express.Request, res: express.Response, next: express.NextFunction) => {
    const start = process.hrtime();
    const id = uuid();
    const reqLog = logger.child({
      req,
      req_id: id,
    });

    res.on('finish', () => {
      const data = {
        res,
        duration: calcDuration(start),
      };
      reqLog.info(data, 'request finish');
    });

    res.on('close', () => {
      const data = {
        res,
        duration: calcDuration(start),

      };
      reqLog.warn(data, 'request socket closed');
    });

    next();

  });

  const collector = new MetricCollector(opts._, {
    logger,
    metricPrefix: opts.metricPrefix,
    redis: opts.url,
    redisPassword: opts.password,
    prefix: opts.prefix,
    autoDiscover: opts.autoDiscover,
  });

  if (opts.autoDiscover) {
    await collector.discoverAll();
  }

  collector.collectJobCompletions();

  app.post('/discover_queues', (_req: express.Request, res: express.Response, next: express.NextFunction) => {
    collector.discoverAll()
      .then(() => {
        res.send({
          ok: true,
        });
      })
      .catch((err: any) => next(err));
  });

  app.get('/healthz', (_req: express.Request, res: express.Response, next: express.NextFunction) => {
    collector.ping()
      .then(() => {
        res.send({
          ok: true,
        });
      })
      .catch((err: any) => next(err));
  });

  app.get('/metrics', (_req: express.Request, res: express.Response, next: express.NextFunction) => {
    collector.updateAll()
      .then(() => {
        res.contentType(promClient.register.contentType);
        res.send(promClient.register.metrics());
      })
      .catch(err => next(err));
  });

  app.use((err: any, _req: express.Request, res: express.Response, _next: express.NextFunction) => {
    res.status(500);
    res.send({
      err: (err && err.message) || 'Unknown error',
    });
  });

  return app;
}

export async function startServer(opts: Options): Promise<{ done: Promise<void> }> {
  const app = await makeServer(opts);

  let server: http.Server;
  await new Promise((resolve, reject) => {
    server = app.listen(opts.port, opts.bindAddress, (err: any) => {
      if (err) {
        reject(err);
        return;
      }
      logger.info(`Running on ${opts.bindAddress}:${opts.port}`);
      resolve();
    });
  });

  process.on('SIGTERM', () => server.close());

  const done = new Promise<void>((resolve, reject) => {
    server.on('close', () => resolve());
    server.on('error', (err: any) => reject(err));
  });

  return { done };
}
