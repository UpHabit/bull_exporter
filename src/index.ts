import promClient from 'prom-client';

import { logger } from './logger';
import { MetricCollector } from './metricCollector';
import { getOptionsFromArgs, Options } from './options';
import { startServer } from './server';

// because we explicitly want just the metrics here
// tslint:disable:no-console

export async function printOnce(opts: Options): Promise<void> {
  const collector = new MetricCollector(opts._, {
    logger,
    metricPrefix: opts.metricPrefix,
    redis: opts.url,
    sentinelExtraEndpoints: opts.sentinelExtraEndpoints,
    sentinelMasterName: opts.sentinelMasterName,
    prefix: opts.prefix,
    autoDiscover: opts.autoDiscover,
  });
  if (opts.autoDiscover) {
    await collector.discoverAll();
  }
  await collector.updateAll();
  await collector.close();

  console.log(promClient.register.metrics());
}

export async function runServer(opts: Options): Promise<void> {
  const { done } = await startServer(opts);
  await done;
}

export async function main(...args: string[]): Promise<void> {
  const opts = getOptionsFromArgs(...args);
  if (opts.once) {
    await printOnce(opts);
  } else {
    await runServer(opts);
  }
}

if (require.main === module) {
  const args = process.argv.slice(2);

  let exitCode = 0;
  main(...args)
    .catch(() => process.exitCode = exitCode = 1)
    .then(() => {
      setTimeout(
        () => {
          logger.error('No clean exit after 5 seconds, force exit');
          process.exit(exitCode);
        },
        5000,
      ).unref();
    })
    .catch(err => {
      console.error('Double error');
      console.error(err.stack);
      process.exit(-1);
    });
}
