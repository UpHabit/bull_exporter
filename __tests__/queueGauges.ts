import { Job } from 'bullmq';
import { getJobCompleteStats, getStats } from '../src/queueGauges';

import { makeQueue, makeWorker, TestData } from './create.util';
import { getCurrentTestHash } from './setup.util';

let testData: TestData;
const JOB_NAME = 'test-job';

beforeEach(async () => {
	jest.resetModules();
	const hash = getCurrentTestHash();
	testData = makeQueue(hash);
});

afterEach(async () => {
	await testData.worker?.close();
	await testData.queue.obliterate({ force: true });
	await testData.events.close();
	await testData.queue.close();
});

it('should list 1 queued job', async () => {
	const { name, queue, prefix, guages, registry } = testData;

	await queue.add(JOB_NAME, { a: 1 });

	await getStats(prefix, name, queue, guages);

	expect(registry.metrics()).toMatchSnapshot();
});

it('should list 1 completed job', async () => {
	const { name, queue, prefix, guages, registry, events } = testData;
	testData.worker = makeWorker(name, async (jobInner: Job<unknown>) => {
		expect(jobInner).toMatchObject({ data: { a: 1 } });
	});

	const job = await queue.add(JOB_NAME, { a: 1 });
	await job.waitUntilFinished(events);

	await getStats(prefix, name, queue, guages);
	await getJobCompleteStats(prefix, name, job, guages);

	expect(registry.metrics()).toMatchSnapshot();
});

it('should list 1 completed job with delay', async () => {
	const { name, queue, prefix, guages, registry, events } = testData;
	testData.worker = makeWorker(name, async (jobInner: Job<unknown>) => {
		expect(jobInner).toMatchObject({ data: { a: 1 } });
	});

	const job = await queue.add(JOB_NAME, { a: 1 });
	await job.waitUntilFinished(events);
	const doneJob: any = await queue.getJob(job.id!);
	// lie about job duration
	doneJob.finishedOn = doneJob.processedOn + 1000;

	await getStats(prefix, name, queue, guages);
	await getJobCompleteStats(prefix, name, doneJob, guages);

	expect(registry.metrics()).toMatchSnapshot();
});

it('should list 1 failed job', async () => {
	const { name, queue, prefix, guages, registry, events } = testData;

	testData.worker = makeWorker(queue.name, async (jobInner: Job<unknown>) => {
		expect(jobInner).toMatchObject({ data: { a: 1 } });
		throw new Error('expected');
	});
	const job = await queue.add(JOB_NAME, { a: 1 });

	await expect(job.waitUntilFinished(events)).rejects.toThrow(/expected/);

	await getStats(prefix, name, queue, guages);

	expect(registry.metrics()).toMatchSnapshot();
});

it('should list 1 delayed job', async () => {
	const { name, queue, prefix, guages, registry, events } = testData;

	const job = await queue.add(JOB_NAME, { a: 1 }, { delay: 100000 });

	await getStats(prefix, name, queue, guages);

	expect(registry.metrics()).toMatchSnapshot();

	// for some reason delayed jobs are not obliterated, so we need to complete it first in order for the cleanup to succeed
	await job.changeDelay(0);
	testData.worker = makeWorker(name, async (jobInner: Job<unknown>) => {
		expect(jobInner).toMatchObject({ data: { a: 1 } });
	});
	await job.waitUntilFinished(events);
});

it('should list 1 active job', async () => {
	const { name, queue, prefix, guages, registry, events } = testData;

	let jobStartedResolve!: () => void;
	let jobDoneResolve!: () => void;
	const jobStartedPromise = new Promise<void>((resolve) => (jobStartedResolve = resolve));
	const jobDonePromise = new Promise<void>((resolve) => (jobDoneResolve = resolve));

	testData.worker = makeWorker(queue.name, async () => {
		jobStartedResolve();
		await jobDonePromise;
	});
	const job = await queue.add(JOB_NAME, { a: 1 });

	await jobStartedPromise;
	await getStats(prefix, name, queue, guages);
	expect(registry.metrics()).toMatchSnapshot();
	jobDoneResolve();
	await job.waitUntilFinished(events);

	await getStats(prefix, name, queue, guages);
	expect(registry.metrics()).toMatchSnapshot();
});
