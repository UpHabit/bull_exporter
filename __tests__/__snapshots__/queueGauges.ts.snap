// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should list 1 active job 1`] = `
"# HELP test_stat_completed Number of completed messages (not yet deleted)
# TYPE test_stat_completed gauge
test_stat_completed{prefix=\\"test-queue\\",queue=\\"2dc07c0ff9d27ef5\\"} 0

# HELP test_stat_complete_duration Time to complete jobs
# TYPE test_stat_complete_duration summary

# HELP test_stat_active Number of active messages
# TYPE test_stat_active gauge
test_stat_active{prefix=\\"test-queue\\",queue=\\"2dc07c0ff9d27ef5\\"} 1

# HELP test_stat_delayed Number of delayed messages
# TYPE test_stat_delayed gauge
test_stat_delayed{prefix=\\"test-queue\\",queue=\\"2dc07c0ff9d27ef5\\"} 0

# HELP test_stat_failed Number of failed messages (not yet deleted)
# TYPE test_stat_failed gauge
test_stat_failed{prefix=\\"test-queue\\",queue=\\"2dc07c0ff9d27ef5\\"} 0

# HELP test_stat_waiting Number of waiting messages
# TYPE test_stat_waiting gauge
test_stat_waiting{prefix=\\"test-queue\\",queue=\\"2dc07c0ff9d27ef5\\"} 0

# HELP test_stat_total_failed Total number of failed messages
# TYPE test_stat_total_failed counter

# HELP test_stat_total_completed Total number of completed messages
# TYPE test_stat_total_completed counter
"
`;

exports[`should list 1 active job 2`] = `
"# HELP test_stat_completed Number of completed messages (not yet deleted)
# TYPE test_stat_completed gauge
test_stat_completed{prefix=\\"test-queue\\",queue=\\"2dc07c0ff9d27ef5\\"} 1

# HELP test_stat_complete_duration Time to complete jobs
# TYPE test_stat_complete_duration summary

# HELP test_stat_active Number of active messages
# TYPE test_stat_active gauge
test_stat_active{prefix=\\"test-queue\\",queue=\\"2dc07c0ff9d27ef5\\"} 0

# HELP test_stat_delayed Number of delayed messages
# TYPE test_stat_delayed gauge
test_stat_delayed{prefix=\\"test-queue\\",queue=\\"2dc07c0ff9d27ef5\\"} 0

# HELP test_stat_failed Number of failed messages (not yet deleted)
# TYPE test_stat_failed gauge
test_stat_failed{prefix=\\"test-queue\\",queue=\\"2dc07c0ff9d27ef5\\"} 0

# HELP test_stat_waiting Number of waiting messages
# TYPE test_stat_waiting gauge
test_stat_waiting{prefix=\\"test-queue\\",queue=\\"2dc07c0ff9d27ef5\\"} 0

# HELP test_stat_total_failed Total number of failed messages
# TYPE test_stat_total_failed counter

# HELP test_stat_total_completed Total number of completed messages
# TYPE test_stat_total_completed counter
"
`;

exports[`should list 1 completed job 1`] = `
"# HELP test_stat_completed Number of completed messages (not yet deleted)
# TYPE test_stat_completed gauge
test_stat_completed{prefix=\\"test-queue\\",queue=\\"7cc276b43bb43656\\"} 1

# HELP test_stat_complete_duration Time to complete jobs
# TYPE test_stat_complete_duration summary

# HELP test_stat_active Number of active messages
# TYPE test_stat_active gauge
test_stat_active{prefix=\\"test-queue\\",queue=\\"7cc276b43bb43656\\"} 0

# HELP test_stat_delayed Number of delayed messages
# TYPE test_stat_delayed gauge
test_stat_delayed{prefix=\\"test-queue\\",queue=\\"7cc276b43bb43656\\"} 0

# HELP test_stat_failed Number of failed messages (not yet deleted)
# TYPE test_stat_failed gauge
test_stat_failed{prefix=\\"test-queue\\",queue=\\"7cc276b43bb43656\\"} 0

# HELP test_stat_waiting Number of waiting messages
# TYPE test_stat_waiting gauge
test_stat_waiting{prefix=\\"test-queue\\",queue=\\"7cc276b43bb43656\\"} 0

# HELP test_stat_total_failed Total number of failed messages
# TYPE test_stat_total_failed counter

# HELP test_stat_total_completed Total number of completed messages
# TYPE test_stat_total_completed counter
test_stat_total_completed{prefix=\\"test-queue\\",queue=\\"7cc276b43bb43656\\"} 1
"
`;

exports[`should list 1 completed job with delay 1`] = `
"# HELP test_stat_completed Number of completed messages (not yet deleted)
# TYPE test_stat_completed gauge
test_stat_completed{prefix=\\"test-queue\\",queue=\\"672d03e07bff9cd3\\"} 1

# HELP test_stat_complete_duration Time to complete jobs
# TYPE test_stat_complete_duration summary
test_stat_complete_duration{quantile=\\"0.01\\",prefix=\\"test-queue\\",queue=\\"672d03e07bff9cd3\\"} 1000
test_stat_complete_duration{quantile=\\"0.05\\",prefix=\\"test-queue\\",queue=\\"672d03e07bff9cd3\\"} 1000
test_stat_complete_duration{quantile=\\"0.5\\",prefix=\\"test-queue\\",queue=\\"672d03e07bff9cd3\\"} 1000
test_stat_complete_duration{quantile=\\"0.9\\",prefix=\\"test-queue\\",queue=\\"672d03e07bff9cd3\\"} 1000
test_stat_complete_duration{quantile=\\"0.95\\",prefix=\\"test-queue\\",queue=\\"672d03e07bff9cd3\\"} 1000
test_stat_complete_duration{quantile=\\"0.99\\",prefix=\\"test-queue\\",queue=\\"672d03e07bff9cd3\\"} 1000
test_stat_complete_duration{quantile=\\"0.999\\",prefix=\\"test-queue\\",queue=\\"672d03e07bff9cd3\\"} 1000
test_stat_complete_duration_sum{prefix=\\"test-queue\\",queue=\\"672d03e07bff9cd3\\"} 1000
test_stat_complete_duration_count{prefix=\\"test-queue\\",queue=\\"672d03e07bff9cd3\\"} 1

# HELP test_stat_active Number of active messages
# TYPE test_stat_active gauge
test_stat_active{prefix=\\"test-queue\\",queue=\\"672d03e07bff9cd3\\"} 0

# HELP test_stat_delayed Number of delayed messages
# TYPE test_stat_delayed gauge
test_stat_delayed{prefix=\\"test-queue\\",queue=\\"672d03e07bff9cd3\\"} 0

# HELP test_stat_failed Number of failed messages (not yet deleted)
# TYPE test_stat_failed gauge
test_stat_failed{prefix=\\"test-queue\\",queue=\\"672d03e07bff9cd3\\"} 0

# HELP test_stat_waiting Number of waiting messages
# TYPE test_stat_waiting gauge
test_stat_waiting{prefix=\\"test-queue\\",queue=\\"672d03e07bff9cd3\\"} 0

# HELP test_stat_total_failed Total number of failed messages
# TYPE test_stat_total_failed counter

# HELP test_stat_total_completed Total number of completed messages
# TYPE test_stat_total_completed counter
"
`;

exports[`should list 1 delayed job 1`] = `
"# HELP test_stat_completed Number of completed messages (not yet deleted)
# TYPE test_stat_completed gauge
test_stat_completed{prefix=\\"test-queue\\",queue=\\"4f2249f79c9d68b5\\"} 0

# HELP test_stat_complete_duration Time to complete jobs
# TYPE test_stat_complete_duration summary

# HELP test_stat_active Number of active messages
# TYPE test_stat_active gauge
test_stat_active{prefix=\\"test-queue\\",queue=\\"4f2249f79c9d68b5\\"} 0

# HELP test_stat_delayed Number of delayed messages
# TYPE test_stat_delayed gauge
test_stat_delayed{prefix=\\"test-queue\\",queue=\\"4f2249f79c9d68b5\\"} 1

# HELP test_stat_failed Number of failed messages (not yet deleted)
# TYPE test_stat_failed gauge
test_stat_failed{prefix=\\"test-queue\\",queue=\\"4f2249f79c9d68b5\\"} 0

# HELP test_stat_waiting Number of waiting messages
# TYPE test_stat_waiting gauge
test_stat_waiting{prefix=\\"test-queue\\",queue=\\"4f2249f79c9d68b5\\"} 0

# HELP test_stat_total_failed Total number of failed messages
# TYPE test_stat_total_failed counter

# HELP test_stat_total_completed Total number of completed messages
# TYPE test_stat_total_completed counter
"
`;

exports[`should list 1 failed job 1`] = `
"# HELP test_stat_completed Number of completed messages (not yet deleted)
# TYPE test_stat_completed gauge
test_stat_completed{prefix=\\"test-queue\\",queue=\\"dd304535bf11c117\\"} 0

# HELP test_stat_complete_duration Time to complete jobs
# TYPE test_stat_complete_duration summary

# HELP test_stat_active Number of active messages
# TYPE test_stat_active gauge
test_stat_active{prefix=\\"test-queue\\",queue=\\"dd304535bf11c117\\"} 0

# HELP test_stat_delayed Number of delayed messages
# TYPE test_stat_delayed gauge
test_stat_delayed{prefix=\\"test-queue\\",queue=\\"dd304535bf11c117\\"} 0

# HELP test_stat_failed Number of failed messages (not yet deleted)
# TYPE test_stat_failed gauge
test_stat_failed{prefix=\\"test-queue\\",queue=\\"dd304535bf11c117\\"} 1

# HELP test_stat_waiting Number of waiting messages
# TYPE test_stat_waiting gauge
test_stat_waiting{prefix=\\"test-queue\\",queue=\\"dd304535bf11c117\\"} 0

# HELP test_stat_total_failed Total number of failed messages
# TYPE test_stat_total_failed counter
test_stat_total_failed{prefix=\\"test-queue\\",queue=\\"dd304535bf11c117\\"} 1

# HELP test_stat_total_completed Total number of completed messages
# TYPE test_stat_total_completed counter
"
`;

exports[`should list 1 queued job 1`] = `
"# HELP test_stat_completed Number of completed messages (not yet deleted)
# TYPE test_stat_completed gauge
test_stat_completed{prefix=\\"test-queue\\",queue=\\"54cceb2dc1d55935\\"} 0

# HELP test_stat_complete_duration Time to complete jobs
# TYPE test_stat_complete_duration summary

# HELP test_stat_active Number of active messages
# TYPE test_stat_active gauge
test_stat_active{prefix=\\"test-queue\\",queue=\\"54cceb2dc1d55935\\"} 0

# HELP test_stat_delayed Number of delayed messages
# TYPE test_stat_delayed gauge
test_stat_delayed{prefix=\\"test-queue\\",queue=\\"54cceb2dc1d55935\\"} 0

# HELP test_stat_failed Number of failed messages (not yet deleted)
# TYPE test_stat_failed gauge
test_stat_failed{prefix=\\"test-queue\\",queue=\\"54cceb2dc1d55935\\"} 0

# HELP test_stat_waiting Number of waiting messages
# TYPE test_stat_waiting gauge
test_stat_waiting{prefix=\\"test-queue\\",queue=\\"54cceb2dc1d55935\\"} 1

# HELP test_stat_total_failed Total number of failed messages
# TYPE test_stat_total_failed counter

# HELP test_stat_total_completed Total number of completed messages
# TYPE test_stat_total_completed counter
"
`;
